queue/abstract/driver/utube.lua

```diff
@@ -141,49 +282,152 @@ function method.touch(self, id, ttr)
     error('utube queue does not support touch')
 end
 
+local function delete_ready(self, id, utube)
+    self.space_ready:delete(id)
+    put_next_ready(self, utube)
+end
+
 -- delete task
 function method.delete(self, id)
+    local transaction_opts = {}
+    if box.cfg.memtx_use_mvcc_engine then
+        transaction_opts = {txn_isolation = 'read-committed'}
+    end
+
+    local commit_requirements = not box.is_in_txn()
+    if commit_requirements then
+        box.begin(transaction_opts)
+    end
+
     local task = self.space:get(id)
     self.space:delete(id)
     if task ~= nil then
+        if self.ready_space then
+            if task[2] == state.TAKEN then
+                put_next_ready(self, task[3])
+            elseif task[2] == state.READY then
+                delete_ready(self, id, task[3])
+            end
+        end
+
         task = task:transform(2, 1, state.DONE)
 
         local neighbour = self.space.index.utube:min{state.READY, task[3]}
+
+        if commit_requirements then
+            box.commit()
+        end
+
         self.on_task_change(task, 'delete')
         if neighbour then
             self.on_task_change(neighbour)
         end
+        return task
+    end
+
+    if commit_requirements then
+        box.commit()
     end
     return task
 end
 
 -- release task
 function method.release(self, id, opts)
+    local transaction_opts = {}
+    if box.cfg.memtx_use_mvcc_engine then
+        transaction_opts = {txn_isolation = 'read-committed'}
+    end
+
+    local commit_requirements = not box.is_in_txn()
+    if commit_requirements then
+        box.begin(transaction_opts)
+    end
+
     local task = self.space:update(id, {{ '=', 2, state.READY }})
     if task ~= nil then
+        if self.ready_space then
+            -- We guarantee that release is called only on TAKEN tasks.
+            local inserted, err =
+                pcall(self.space_ready.insert, self.space_ready, {id, task[3]})
```

Comment:
In which cases we are expected to fail here?

Reply:
Yes, we guarantee that release is called only on TAKEN tasks, so this task_id and this utube could not be in the space_ready. This means that there could be no errors in a standard work flow. But, for example, user can call `release` method on an already `READY` task. So we need a protection in case of incorrect usage of the tube.

Reply:
> user can call release method on an already `READY` task
So it means you don't actually "guarantee that release is called only on `TAKEN` tasks", you expect it to be so. And isn't it easier to just call "replace" in this case?

Reply:
`space_ready:replace` will break the logic. We need to try to insert and fail on the error. Another ability here:

But it seems we need to update the `_ready` space on the error case + if task_id < task_id in the ready space.

Reply:
Well, ok. Let's update the comment at least